File: app/src/main/java/sugtao4423/lod/main_fragment/Fragment_mention.java
package sugtao4423.lod.main_fragment;

import twitter4j.Paging;
import twitter4j.ResponseList;
import twitter4j.Status;
import twitter4j.TwitterException;

import android.os.AsyncTask;
import android.os.Bundle;
import android.os.Handler;
import android.support.v4.app.Fragment;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v4.widget.SwipeRefreshLayout.OnRefreshListener;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import sugtao4423.lod.App;
import sugtao4423.lod.ListViewListener;
import sugtao4423.lod.R;
import sugtao4423.lod.ShowToast;
import sugtao4423.lod.tweetlistview.EndlessScrollListener;
import sugtao4423.lod.tweetlistview.TweetListAdapter;
import sugtao4423.lod.tweetlistview.TweetListView;

public class Fragment_mention extends Fragment{

	private TweetListView list;
	private SwipeRefreshLayout pulltoRefresh;
	private TweetListAdapter adapter;
	private Handler handler;
	private App app;

	@Override
	public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState){
		app = (App)container.getContext().getApplicationContext();
		handler = new Handler();
		View v = inflater.inflate(R.layout.fragment_list, container, false);
		list = (TweetListView)v.findViewById(R.id.listLine);

		adapter = new TweetListAdapter(container.getContext());
		adapter.setOnItemClickListener(new ListViewListener());
		adapter.setOnItemLongClickListener(new ListViewListener());
		list.setAdapter(adapter);

		final EndlessScrollListener scrollListener = getLoadMoreListener();
		list.addOnScrollListener(scrollListener);

		pulltoRefresh = (SwipeRefreshLayout)v.findViewById(R.id.ListPull);
		pulltoRefresh.setColorSchemeResources(android.R.color.holo_blue_bright, android.R.color.holo_green_light,
				android.R.color.holo_orange_light, android.R.color.holo_red_light);
		OnRefreshListener onRefreshListener = new OnRefreshListener(){

			@Override
			public void onRefresh(){
				adapter.clear();
				loadMention();
				scrollListener.resetState();
			}
		};
		pulltoRefresh.setOnRefreshListener(onRefreshListener);
		onRefreshListener.onRefresh();
		return v;
	}

	public EndlessScrollListener getLoadMoreListener(){
		return new EndlessScrollListener(list.getLinearLayoutManager()){

			@Override
			public void onLoadMore(int current_page){
				if(adapter.getItemCount() > 30)
					loadMention();
			}
		};
	}

	public void loadMention(){
		new AsyncTask<Void, Void, ResponseList<Status>>(){
			@Override
			protected ResponseList<twitter4j.Status> doInBackground(Void... params){
				try{
					if(adapter.getItemCount() > 0){
						long tweetId = adapter.getItem(adapter.getItemCount() - 1).getId();
						return app.getTwitter().getMentionsTimeline(new Paging(1, 50).maxId(tweetId - 1));
					}else{
						return app.getTwitter().getMentionsTimeline(new Paging(1, 50));
					}
				}catch(TwitterException e){
					return null;
				}
			}

			@Override
			protected void onPostExecute(ResponseList<twitter4j.Status> result){
				if(result != null)
					addAll(result);
				else
					new ShowToast(getContext().getApplicationContext(), R.string.error_getMention);
				pulltoRefresh.setRefreshing(false);
				pulltoRefresh.setEnabled(true);
			}
		}.execute();
	}

	public void insert(Status status){
		adapter.insertTop(status);
		handler.post(new Runnable(){

			@Override
			public void run(){
				if(list.getLinearLayoutManager().findFirstVisibleItemPosition() <= 1)
					list.smoothScrollToPosition(0);
			}
		});
	}

	public void addAll(ResponseList<Status> status){
		adapter.addAll(status);
	}
}

File: app/src/main/java/sugtao4423/lod/main_fragment/Fragment_mention.kt
package sugtao4423.lod.main_fragment

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import sugtao4423.lod.App
import sugtao4423.lod.R
import sugtao4423.lod.ShowToast
import sugtao4423.lod.databinding.FragmentListBinding
import sugtao4423.lod.tweetlistview.EndlessScrollListener
import sugtao4423.lod.tweetlistview.TweetListAdapter
import twitter4j.Paging
import twitter4j.ResponseList
import twitter4j.Status
import twitter4j.TwitterException

class Fragment_mention : Fragment() {

    private lateinit var binding: FragmentListBinding
    private lateinit var adapter: TweetListAdapter
    private lateinit var app: App

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        binding = FragmentListBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        app = requireContext().applicationContext as App

        adapter = TweetListAdapter(requireContext())
        binding.listLine.adapter = adapter

        val scrollListener = getLoadMoreListener()
        binding.listLine.addOnScrollListener(scrollListener)

        binding.listPull2Refresh.setColorSchemeResources(android.R.color.holo_blue_bright, android.R.color.holo_green_light,
                android.R.color.holo_orange_light, android.R.color.holo_red_light)
        val onRefreshListener = SwipeRefreshLayout.OnRefreshListener {
            adapter.clear()
            loadMention()
            scrollListener.resetState()
        }
        binding.listPull2Refresh.setOnRefreshListener(onRefreshListener)
        onRefreshListener.onRefresh()
    }

    private fun getLoadMoreListener(): EndlessScrollListener {
        return object : EndlessScrollListener(binding.listLine.linearLayoutManager) {
            override fun onLoadMore(currentPage: Int) {
                if (adapter.hasNextPage) {
                    loadMention()
                }
            }
        }
    }

    private fun loadMention() {
        CoroutineScope(Dispatchers.Main).launch {
            val result = withContext(Dispatchers.IO) {
                val paging = Paging(1, 50).let {
                    if (adapter.itemCount > 0) it.maxId(adapter.data.last().id - 1) else it
                }

                try {
                    app.getTwitter().getMentionsTimeline(paging)
                } catch (e: TwitterException) {
                    null
                }
            }
            if (result != null) {
                if (result.isEmpty()) {
                    adapter.hasNextPage = false
                }
                addAll(result)
            } else {
                ShowToast(requireContext().applicationContext, R.string.error_get_mention)
            }
            binding.listPull2Refresh.isRefreshing = false
            binding.listPull2Refresh.isEnabled = true
        }
    }

    fun insert(status: Status) {
        adapter.insertTop(status)
        if (binding.listLine.linearLayoutManager.findFirstVisibleItemPosition() <= 1) {
            binding.listLine.smoothScrollToPosition(0)
        }
    }

    private fun addAll(statuses: ResponseList<Status>) {
        adapter.addAll(statuses)
    }

}
